util/webhook/webhook.go

```diff
@@ -160,40 +182,95 @@ func (a *ArgoCDWebhookHandler) HandleEvent(payload interface{}) {
 	}
 
 	for _, webURL := range webURLs {
-		urlObj, err := url.Parse(webURL)
-		if err != nil {
-			log.Warnf("Failed to parse repoURL '%s'", webURL)
-			continue
-		}
-		regexpStr := `(?i)(http://|https://|\w+@|ssh://(\w+@)?)` + urlObj.Host + "(:[0-9]+|)[:/]" + urlObj.Path[1:] + "(\\.git)?"
-		repoRegexp, err := regexp.Compile(regexpStr)
-		if err != nil {
-			log.Warnf("Failed to compile regexp for repoURL '%s'", webURL)
-			continue
-		}
-
 		for _, app := range apps.Items {
-			if !repoRegexp.MatchString(app.Spec.Source.RepoURL) {
-				log.Debugf("%s does not match", app.Spec.Source.RepoURL)
-				continue
-			}
-			targetRev := app.Spec.Source.TargetRevision
-			if targetRev == "HEAD" || targetRev == "" {
-				if !touchedHead {
+			if appRevisionHasChanged(&app, revision, touchedHead) && appUsesURL(&app, webURL) && appFilesHaveChanged(&app, changedFiles) {
+				_, err = argo.RefreshApp(appIf, app.ObjectMeta.Name, v1alpha1.RefreshTypeNormal)
+				if err != nil {
+					log.Warnf("Failed to refresh app '%s' for controller reprocessing: %v", app.ObjectMeta.Name, err)
 					continue
 				}
-			} else if targetRev != revision {
-				continue
-			}
-			_, err = argo.RefreshApp(appIf, app.ObjectMeta.Name, v1alpha1.RefreshTypeNormal)
-			if err != nil {
-				log.Warnf("Failed to refresh app '%s' for controller reprocessing: %v", app.ObjectMeta.Name, err)
-				continue
 			}
 		}
 	}
 }
 
+func getAppRefreshPrefix(app *v1alpha1.Application) string {
+	// an explicitPrefix setting always takes precence
+	if explicitPrefix := app.Annotations["argocd.argoproj.io/refresh-prefix"]; explicitPrefix != "" {
```

Comment:
I would define all annotations in common.go as constants and document each of them in const's comments.